internal_types/src/schema/merge.rs

```diff
@@ -0,0 +1,447 @@
+        field_name,
+        existing_data_type,
+        new_data_type
+    ))]
+    TryMergeBadArrowType {
+        field_name: String,
+        existing_data_type: ArrowDataType,
+        new_data_type: ArrowDataType,
+    },
+
+    #[snafu(display(
+        "Schema Merge Error: Incompatible nullability for '{}'. Existing field {}, new field {}",
+        field_name, nullable_to_str(*existing_nullability), nullable_to_str(*new_nullability)
+    ))]
+    TryMergeBadNullability {
+        field_name: String,
+        existing_nullability: bool,
+        new_nullability: bool,
+    },
+}
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+fn nullable_to_str(nullability: bool) -> &'static str {
+    if nullability {
+        "can be null"
+    } else {
+        "can not be null"
+    }
+}
+
+/// Schema Merger
+///
+/// The usecase for merging schemas is when different chunks have
+/// different schemas. This struct can be used to build a combined
+/// schema by merging Schemas together according to the following
+/// rules:
+///
+/// 1. New columns may be added in subsequent schema, but the types of
+///    the columns (including any metadata) must be the same
+///
+/// 2. The measurement names must be consistent: one or both can be
+///    `None`, or they can both be `Some(name`)
+#[derive(Debug, Default)]
+pub struct SchemaMerger {
+    fields: HashMap<String, (Field, Option<InfluxColumnType>)>,
+    measurement: Option<String>,
+}
+
+impl SchemaMerger {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Appends the schema to the merged schema being built,
+    /// validating that no columns are added.
+    ///
+    /// O(n) in the number of fields (columns)
+    pub fn merge(&mut self, other: &Schema) -> Result<&mut Self> {
+        // Verify measurement name is compatible
+        match (self.measurement.as_ref(), other.measurement()) {
+            (Some(existing_measurement), Some(new_measurement)) => {
+                if existing_measurement != new_measurement {
+                    return TryMergeDifferentMeasurementNames {
+                        existing_measurement,
+                        new_measurement,
+                    }
+                    .fail();
+                }
+            }
+            (None, Some(other)) => self.measurement = Some(other.clone()),
+            _ => {}
+        }
+
+        // Merge fields
+        for (column_type, field) in other.iter() {
+            self.merge_field(field, column_type)?;
+        }
+
+        Ok(self)
+    }
+
+    fn merge_field(
+        &mut self,
+        field: &Field,
+        column_type: Option<InfluxColumnType>,
+    ) -> Result<&mut Self> {
+        let field_name = field.name();
+        match self.fields.raw_entry_mut().from_key(field_name) {
+            RawEntryMut::Vacant(vacant) => {
+                // Purposefully don't propagate metadata
```

Comment:
```suggestion
                // Purposefully don't propagate metadata
                // to avoid propagating information such as sort key
```
I think comments are best  when used to communicate the "why" which can not always be gleaned from the code or tests.